File: tests/TestIMU.cpp
#include "MBInterface.hpp"
#include "IMUInterface.hpp"

#include <chrono>
#include <thread>

int main()
{
  Roki::MBInterface mbi;
  Roki::MBInterface::TTYConfig tty;

  tty.Port = "/dev/ttyAMA2";
  tty.Baudrate = 9600;
  tty.ParityBit = true;
  tty.Stopbits = tty.STOPBITS_ONE;
  tty.Timeout = 2;

  if (!mbi.Configure(tty))
  {
    std::cout << mbi.GetError() << std::endl;
    exit(1);
  }

  Roki::IMUInterface imu;
  Roki::IMUInterface::IMUFrame frame;
  Roki::IMUInterface::IMUInfo info;

  uint8_t fr[5]{0, 1, 2, 3, 4};
  int i = -1;
  int j = 0;

  while (false)
  {
    if (j++ == 5)
      if (!imu.ResetCounter(mbi))
      {
        std::cout << imu.GetError() << std::endl;
        std::this_thread::sleep_for(std::chrono::seconds(1));
        continue;
      }

    if (!imu.GetInfo(mbi, info))
    {
      std::cout << imu.GetError() << std::endl;
      std::this_thread::sleep_for(std::chrono::seconds(1));
      continue;
    }

    std::cout << "Imu info: " << std::endl;

    std::cout << "  First:  " << info.First << std::endl;
    std::cout << "  NumAv:  " << info.NumAv << std::endl;

    std::cout << std::endl;
  }

  while (true)
  {
    i = (i + 1) % 500;
    if (!imu.GetFrame(mbi, i, frame))
    {
      std::cout << imu.GetError() << std::endl;
      std::this_thread::sleep_for(std::chrono::seconds(1));
      continue;
    }

    std::cout << "Quaternion: " << +i << std::endl;
    std::cout << "  x:  " << frame.Orientation.X << std::endl;
    std::cout << "  y:  " << frame.Orientation.Y << std::endl;
    std::cout << "  z:  " << frame.Orientation.Z << std::endl;
    std::cout << "  w:  " << frame.Orientation.W << std::endl;
    // std::cout << "  Acc: " << frame.Orientation.Accuracy << std::endl;

    std::cout << "Timestamp: " << std::endl;
    std::cout << "  s:  " << frame.Timestamp.TimeS << std::endl;
    std::cout << "  ns: " << frame.Timestamp.TimeNS << std::endl;

    std::cout << "Sensor ID: " << +frame.SensorID << std::endl;
    std::this_thread::sleep_for(std::chrono::seconds(1));
  }
  return 0;
}

File: tests/scripts/TestImu.py
import sys
sys.path.append('/home/tndrd/starkit/kondo_2023/roki-mb-interface/build')

import Roki

mb = Roki.Motherboard()

conf = Roki.TTYConfig()

conf.Port = "/dev/ttyAMA2"
conf.Baudrate = 1250000
conf.Stopbits = Roki.Stopbits.One
conf.ParityBit = True
conf.Timeout = 2

mb.Configure(conf)

while True:
    fr = mb.GetCurrentOrientation()

    print("Quaternion: ")
    print("  X:  " + fr.Orientation.X)
    print("  Y:  " + fr.Orientation.Y)
    print("  Z:  " + fr.Orientation.Z)
    print("  W:  " + fr.Orientation.W)
    print("Timestamp: ")
    print("  S:  " + fr.Timestamp.TimeS)
    print("  NS: " + fr.Timestamp.TimeNS)
    print("SensorID: " + fr.SensorID)
    print("\n")

